app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt

```diff
@@ -34,7 +38,31 @@ open class FeedViewModel : ViewModel() {
     }
 
     fun refresh() {
-        //todo: fix the implementation
+        apiServiceDisposable = FeedApiService.feedApiService.getFeed()
+            .subscribeOn(Schedulers.io())
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribe({ feedResponse ->
+                updateState {  State(feedResponse.templatesMetadata.toFeedItems(),
+                                getState().isLoading) }
+            },{ error ->
+                networkErrorEvent.postValue(Event(error.message.toString()))
+            })
+    }
```

Comment:
1) Check that we don't have already a task in progress - if (getState.isLoading) return
2) Update the is loading flag at the beginning before we are executing the task and turn it off on terminal event
3) Dispose the previous task - although after the check I mentioned in 1) it should does nothing
4) The way we should use the update state method is by using the copy constructor instead of manually copying the other values.

Reply:
I didnt understand 3.
if at the begining i am checking if there is allready a refresh task running so I neeed to return so if I will dispode the runnung refresh so I would end up not refreshing at all.